### ISSUE TITLE
ClientWebSocket.CloseAsync hangs if internet connection is lost

### ISSUE THREAD

**Issue by ghost** [dotnet/runtime#51590]
We have an app that connects to a websocket service with a custom "heartbeat" mechanism. On an interval, the client requests a heartbeat from the server (via websocket message), and if the server doesn't reply, the client assumes the server is gone, and reconnects.

We have essentially the following code:
```cs
if (!hasReceivedHeartbeatAck)
{
    logger.Warning("Did not receive heartbeat Ack from gateway");
    await client.Disconnect(WebSocketCloseStatus.ProtocolError, "Did not receive ACK in time", default);
    logger.Information("Connection closed, reconnecting...");
    await Reconnect();
}
```

(the original code is [here](https://github.com/xSke/PluralKit/blob/a2d2036851139c18736371600f1d2625f3b2ba62/Myriad/Gateway/Shard.cs#L127), it calls [this method](https://github.com/xSke/PluralKit/blob/a2d2036851139c18736371600f1d2625f3b2ba62/Myriad/Gateway/ShardConnection.cs#L127) then the `return false` propagates up a couple methods and eventually gets to [this log line](https://github.com/xSke/PluralKit/blob/a2d2036851139c18736371600f1d2625f3b2ba62/Myriad/Gateway/Shard.cs#L90))

When there's a server error, this works fine - the connection is correctly closed and reconnected.
However, if the internet connection is lost (this can be reproduced by disconnecting the ethernet wire), `client.Disconnect` **hangs until the internet connection returns**, then throws the following error:
> System.Net.WebSockets.WebSocketException (2): The remote party closed the WebSocket connection without completing the close handshake. ---> System.IO.IOException: Unable to read data from the transport connection: Connection reset by peer. ---> System.Net.Sockets.SocketException (104): Connection reset by peer

**Comment by ManickaP** [comment]
Triage: we should investigate possible solutions to the hang, like TCP_KEEPALIVE.

**Comment by CarnaViire** [comment]
Triage:

_**TL;DR:** Use a `CancellationToken` to avoid hanging in `CloseAsync`. Alternatively, set up TCP_KEEPALIVE to enable detection on the underlying transport level._

`ClientWebSocket.CloseAsync` method accepts a `CancellationToken` that can be used to avoid waiting longer than the user deems necessary.

Unless `TCP_KEEPALIVE` is set up on the TCP connection, neither `WebSocket` nor underlying `NetworkStream` have no way of knowing whether the connection got lost or the peer just isn't answering. The users are advised to handle the second case (peer technically alive, but very slow) anyway, either by imposing a timeout with a `CancellationToken` or by configuring TCP_KEEPALIVE on the TCP socket (e.g. via [`SocketsHttpHandler.ConnectCallback`](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler.connectcallback?view=net-9.0#examples))

**NOTE:**
The [PING-PONG KeepAlive mechanism](https://learn.microsoft.com/en-us/dotnet/fundamentals/networking/websockets#keep-alive-strategies) (that allows the users to detect stale connections) is **NOT applicable** to `CloseAsync`. This mechanism requires a PING frame to be sent, but as soon as the close handshake is initiated by sending a CLOSE frame, by protocol, no other frames could be sent, including PING.

Closing as answered.
